src/watchdog.js

```diff
@@ -0,0 +1,414 @@
+ *
+ * Other important APIs:
+ *
+ *		watchdog.on( 'error', () => { console.log( 'Editor crashed.' ) } );
+ *		watchdog.on( 'restart', () => { console.log( 'Editor was restarted.' ) } );
+ *
+ *		watchdog.restart(); // Restarts the editor.
+ *		watchdog.destroy(); // Destroys the editor and global error event listeners.
+ *
+ * 		watchdog.editor; // Current editor instance.
+ *
+ * 		watchdog.crashes.forEach( crashInfo => console.log( crashInfo ) );
+ */
+export default class Watchdog {
+	/**
+	 * @param {Object} config The watchdog plugin configuration.
+	 * @param {Number} config.crashNumberLimit A threshold specifying the number of crashes
+	 * when the watchdog stops restarting the editor in case of errors.
+	 * @param {Number} config.waitingTime A minimum amount of time between saving editor data internally.
+	 */
+	constructor( { crashNumberLimit, waitingTime } = {} ) {
+		/**
+		 * An array of crashes saved as an object with the following props:
+		 * * message: String,
+		 * * source: String,
+		 * * lineno: String,
+		 * * colno: String
+		 *
+		 * @public
+		 * @readonly
+		 * @type {Array.<Object>}
+		 */
+		this.crashes = [];
+
+		/**
+		 * Crash number limit (default to 3). After the limit is reached the editor is not restarted by the watchdog.
+		 * This is to prevent an infinite crash loop.
+		 *
+		 * @private
+		 * @type {Number}
+		 */
+		this._crashNumberLimit = crashNumberLimit || 3;
+
+		/**
+		 * Checks if the event error comes from the editor that is handled by the watchdog (by checking the error context)
+		 * and restarts the editor.
+		 *
+		 * @private
+		 * @type {Function}
+		 */
+		this._boundErrorHandler = this._handleGlobalErrorEvent.bind( this );
+
+		/**
+		 * Throttled save method. The `save()` method is called the specified `waitingTime` after `throttledSave()` is called,
+		 * unless a new action happens in the meantime.
+		 *
+		 * @private
+		 * @type {Function}
+		 */
+		this._throttledSave = throttle( this._save.bind( this ), waitingTime || 5000 );
+
+		/**
+		 * The current editor.
+		 *
+		 * @private
+		 * @type {module:core/editor/editor~Editor}
+		 */
+		this._editor = null;
+
+		/**
+		 * The editor creation method.
+		 *
+		 * @private
+		 * @member {Function} #_creator
+		 * @see #setCreator
+		 */
+
+		/**
+		 * The editor destruction method.
+		 *
+		 * @private
+		 * @member {Function} #_destructor
+		 * @see #setDestructor
+		 */
+
+		/**
+		 * The latest saved editor data.
+		 *
+		 * @private
+		 * @member {String} #_data
+		 */
+
+		/**
+		 * The last document version.
+		 *
+		 * @private
+		 * @member {Number} #_lastDocumentVersion
+		 */
+
+		/**
+		* The editor source element or data.
+		*
+		* @private
+		* @member {HTMLElement|String} #_elementOrData
+		*/
+
+		/**
+		* The editor configuration.
+		*
+		* @private
+		* @member {Object|undefined} #_config
+		*/
+	}
+
+	/**
+	 * The current editor instance.
+	 *
+	 * @type {module:core/editor/editor~Editor}
```

Comment:
Not sure if we need/should document this as `@readonly` if it has only a getter anyway? How does our docs builder handle it?